livre/core/pipeline/AsyncData.cpp

```diff
@@ -0,0 +1,171 @@
+/* Copyright (c) 2011-2016, EPFL/Blue Brain Project
+ *                     Ahmet Bilgili <ahmet.bilgili@epfl.ch>
+ *
+ * This file is part of Livre <https://github.com/BlueBrain/Livre>
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License version 3.0 as published
+ * by the Free Software Foundation.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include <livre/core/pipeline/AsyncData.h>
+#include <livre/core/pipeline/PortData.h>
+#include <livre/core/pipeline/Future.h>
+
+#include <boost/thread/future.hpp>
+
+namespace livre
+{
+
+typedef boost::shared_future< PortDataPtr > ConstPortDataFuture;
+typedef boost::promise< PortDataPtr > ConstPortDataPromise;
+typedef std::vector< ConstPortDataFuture > ConstPortDataFutures;
+
+namespace
+{
+    /* Boost wait_for_any implementation uses the internal locks from the futures.
+       Therefore, the get/set/query operations on promises/futures causes deadlocks
+       when used with wait_for_any( futurelist ) if futurelist is including the
+       future to be queried. With below implementation no locking is needed between
+       future operations. A higher granularity can be added for checking whether
+       the list includes the future or not, but simply exiting the wait_for_any
+       operation when lock is owned is simpler.
+     */
+    ReadWriteMutex waitForAnyLock;
+}
+
+struct AsyncData::Impl
+{
+    Impl( const DataInfo& dataInfo )
+        : _future( _promise.get_future( ))
+        , _dataInfo( dataInfo )
+    {}
+
+    ~Impl()
+    {
+        if( !isReady())
+            _promise.set_value( PortDataPtr( ));
+    }
+
+    const PortDataPtr& get( const std::type_index& dataType ) const
+    {
+        ReadLock lock( waitForAnyLock );
+
+        const PortDataPtr& data = _future.get();
```

Comment:
bind to reference only if you're certain that boost::shared_ptr is holding a strong reference of the returned type, otherwise you're returning a reference to a temporary that the compiler won't detect as bad.
If you want to stay safe, remove the reference.